src/compiler/checker.ts

```diff
@@ -24121,6 +24111,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const templateType = getTemplateTypeFromMappedType(target);
         const inference = createInferenceInfo(typeParameter);
         inferTypes([inference], sourceType, templateType);
+        const sourceValueDeclaration = sourceType.symbol?.valueDeclaration;
+        if (sourceValueDeclaration) {

```

Comment:
I expect that this part is not fully correct. What I'd like to do here is to only get `sourceValueDeclaration` if it's an anonymous/fresh/smth declaration **within** within the argument position. I'm not sure how to check for this.

Reply:
Use `findAncestor` to check for argument position membership, use `ObjectFlags.FreshLiteral` to check for freshness on the type.

Reply:
But doesn't this all just imply we should store intra-expression inference sites in a per-property map or something?

Reply:
I added `.valueDeclaration` to reverse mapped properties. This grants me clean access to the node I'm looking for and fixes an issue with array literals used as values of object reverse mapped types ([test case](https://github.com/microsoft/TypeScript/pull/54029/commits/fae5ae6e1b65c7c00be0424fba7a961ea8247f9e)). I think that this gives me now exactly what I'm looking for when combined with the freshness check and I don't need to do any `findAncestor` traversal.